pennylane/transforms/optimization/pattern_match.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Transform for simple template optimization."""
+
+from pennylane import apply
+from pennylane.tape import get_active_tape
+from pennylane.wires import Wires
+from pennylane.transforms import qfunc_transform
+import pennylane.ops.qubit as ops
+
+from .optimization_utils import find_next_gate
+
+# Instead of using the circuit DAG, I tried doing this with string matching.
+# It works, but its maybe not the best solution.
+templates = {
+    2: {"S0 S0": [(ops.PauliZ, 0)], "T0 T0": [(ops.S, 0)],},
+    3: {
+        "Hadamard0 PauliX0 Hadamard0": [(ops.PauliZ, 0)],
+        "Hadamard0 PauliZ0 Hadamard0": [(ops.PauliX, 0)],
+        "CNOT01 PauliX0 CNOT01": [(ops.PauliX, 0), (ops.PauliX, 1)],
+        "CNOT01 Toffoli012 CNOT01": [(ops.Toffoli, [0, 1, 2]), (ops.CNOT, [0, 2])],
+    },
+    4: {"CNOT01 Hadamard0 Hadamard1 CNOT10": [(ops.Hadamard, 0), (ops.Hadamard, 1)]},
```

Comment:
Q: how should we handle cases that differ by commuting gates? For example,

```
"CNOT01 Hadamard1 Hadamard0 CNOT10"
```

If the templates we're working with are small and there aren't a lot of permutations, it might be fastest to just have both versions in the list. Alternatively, we implement more general commutation rules to order gates commuting gates in a canonical order / wire order so that by default the patterns will be of this form.